packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+      }
+      return (trimmedAccounts, trimmedIssuedOns, trimmedSigners);
+    } else {
+      return (accounts, issuedOns, signers);
+    }
+  }
+
+  /**
+   * @notice Returns identifiers mapped (by unrevoked signers) to a
+   *   given account address by a list of trusted issuers
+   * @param account Address of the account
+   * @param trustedIssuers Array of n issuers whose identifier mappings will be used
+   * @param maxIdentifiers Limit the number of identifiers that will be returned
+   * @return Array (length <= maxIdentifiers) of identifiers
+   * @dev Adds identifier info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+
+  // TODO ASv2 consider also returning an array with counts of identifiers per issuer
+
+  function lookupIdentifiersByAddress(
+    address account,
+    address[] memory trustedIssuers,
+    uint256 maxIdentifiers
+  ) public view returns (bytes32[] memory) {
+    // Same as for the other lookup, preallocate and then trim for now
+    uint256 currIndex = 0;
+    bytes32[] memory identifiers = new bytes32[](maxIdentifiers);
+
+    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
+      address trustedIssuer = trustedIssuers[i];
+      uint256 numIssuersForAddress = addressToIdentifiers[account][trustedIssuer].length;
+      for (uint256 j = 0; j < numIssuersForAddress; j = j.add(1)) {
+        // Iterate through the list of identifiers
+        if (currIndex < maxIdentifiers) {
+          bytes32 identifier = addressToIdentifiers[account][trustedIssuer][j];
+          // Check if the mapping was produced by a revoked signer
+          // solhint-disable-next-line max-line-length
+          OwnershipAttestation[] memory attestationsForIssuer = identifierToAddresses[identifier][trustedIssuer];
+          for (uint256 k = 0; k < attestationsForIssuer.length; k = k.add(1)) {
+            OwnershipAttestation memory attestation = attestationsForIssuer[k];
+            // (identifier, account, issuer) tuples should be unique
+            if (attestation.account == account && !revokedSigners[attestation.signer]) {
+              identifiers[currIndex] = identifier;
+              currIndex = currIndex.add(1);
+              break;
+            }
+          }
+        } else {
+          break;
+        }
+      }
+    }
+    if (currIndex < maxIdentifiers) {
+      // Allocate and fill properly-sized array
+      bytes32[] memory trimmedIdentifiers = new bytes32[](currIndex);
+      for (uint256 i = 0; i < currIndex; i = i.add(1)) {
+        trimmedIdentifiers[i] = identifiers[i];
+      }
+      return trimmedIdentifiers;
+    } else {
+      return identifiers;
+    }
+  }
+
+  // TODO do we want to restrict permissions, or should anyone
+  // with a valid signature be able to register an attestation?
+  modifier isValidUser(address issuer, address account) {
+    require(
+      msg.sender == account ||
```

Comment:
We likely want to enable user accounts to assign a signer key to manage attestations for them. Probably through a new CIP10 path.

Reply:
I may be misunderstanding your comment here, but we have decided to go with the existing `AttestationSigner` role for the time being because it offers the advantages of other CIP10 roles and also has existing CLI/SDK tooling for interacting and setting these roles. Do you think there's a compelling reason to create an entirely new role?

Reply:
Oh, let me clarify that. What I meant is that right now issuers can sign using their account key or a signer key. What I was suggesting is that client accounts also be able to do this. Thinking about it again though, I don't think this is a particularly beneficial idea because this is checking the seder, and it would be unusual for a client to send a transaction from an address that is not their wallet, partly because it would not be funded.